cybozu/tcp.hpp

```diff
@@ -149,21 +154,65 @@ class tcp_socket: public resource {
     //
     // @return `true` if this socket is valid, `false` otherwise.
     bool sendv_close(const iovec* iov, int iovcnt) {
-        if( iovcnt >= MAX_IOVCNT )
-            throw std::logic_error("<tcp_socket::sendv> too many iov.");
-        lock_guard g(m_lock);
-        if( ! _sendv(iov, iovcnt, g) )
-            return false;
-        m_shutdown = true;
-        if( empty() ) {
-            _flush();
+        return with_fd([=](int fd) -> bool {
+            if( iovcnt >= MAX_IOVCNT )
+                throw std::logic_error("<tcp_socket::sendv> too many iov.");
+            lock_guard g(m_lock);
+            if( ! _sendv(fd, iov, iovcnt, g) )
+                return false;
+            m_shutdown = true;
+            if( empty() ) {
+                _flush(fd);
+                return false;
```

Comment:
Original code returns true when `empty()`, but new code returns false. Why is there this difference?

Reply:
Because this is `sendv_close`.
This function has to close the socket after a successful `send` call,
and `with_fd` invalidates and (eventually) closes the socket if the
callback function returns `false`, this should return `false` here.

Reply:
I understand.